genesis/utils/array_class.py

```diff
@@ -356,19 +358,21 @@ def get_diff_contact_input(solver, max_contacts_per_pair):
     _B = solver._B
 
     return StructDiffContactInput(
-        geom_a=V(dtype=gs.ti_int, shape=(_B, max_contacts_per_pair)),
-        geom_b=V(dtype=gs.ti_int, shape=(_B, max_contacts_per_pair)),
-        local_pos1_a=V_VEC(3, dtype=gs.ti_float, shape=(_B, max_contacts_per_pair)),
-        local_pos1_b=V_VEC(3, dtype=gs.ti_float, shape=(_B, max_contacts_per_pair)),
-        local_pos1_c=V_VEC(3, dtype=gs.ti_float, shape=(_B, max_contacts_per_pair)),
-        local_pos2_a=V_VEC(3, dtype=gs.ti_float, shape=(_B, max_contacts_per_pair)),
-        local_pos2_b=V_VEC(3, dtype=gs.ti_float, shape=(_B, max_contacts_per_pair)),
-        local_pos2_c=V_VEC(3, dtype=gs.ti_float, shape=(_B, max_contacts_per_pair)),
-        w_local_pos1=V_VEC(3, dtype=gs.ti_float, shape=(_B, max_contacts_per_pair)),
-        w_local_pos2=V_VEC(3, dtype=gs.ti_float, shape=(_B, max_contacts_per_pair)),
-        ref_id=V(dtype=gs.ti_int, shape=(_B, max_contacts_per_pair)),
-        valid=V(dtype=gs.ti_int, shape=(_B, max_contacts_per_pair)),
-        ref_penetration=V(dtype=gs.ti_float, shape=(_B, max_contacts_per_pair), needs_grad=True),
+        geom_a=V(dtype=gs.ti_int, shape=_s((_B, max_contacts_per_pair), solver._requires_grad)),
```

Comment:
Define `requires_grad = solver._requires_grad` to reduce line width. It would be more readable.

Reply:
done